forge-gui/res/cardsfolder/h/henzie_toolbox_torre.txt

```diff
@@ -2,7 +2,7 @@ Name:Henzie "Toolbox" Torre
 ManaCost:B R G
 Types: Legendary Creature Devil Rogue
 PT:3/3
-S:Mode$ Continuous | Affected$ Creature.YouCtrl | AffectedZone$ Stack | AddKeyword$ Blitz:CardManaCost:Spell.Creature+cmcGE4 | Description$ Each creature spell you cast with mana value 4 or greater has blitz. The blitz cost is equal to its mana cost. (You may choose to cast that spell for its blitz cost. If you do, it gains haste and "When this creature dies, draw a card." Sacrifice it at the beginning of the next end step.)
+S:Mode$ Continuous | Affected$ Creature.YouCtrl+cmcGE4 | AffectedZone$ Stack | AddKeyword$ Blitz:CardManaCost:Spell.Creature | Description$ Each creature spell you cast with mana value 4 or greater has blitz. The blitz cost is equal to its mana cost. (You may choose to cast that spell for its blitz cost. If you do, it gains haste and "When this creature dies, draw a card." Sacrifice it at the beginning of the next end step.)
```

Comment:
The previous implementation was so close to being correct!

Reply:
This Check is for MDFC

is the affected checked again when casting the Backside?

Reply:
I'll look into this later today (it's currently 3am local time for me).

Reply:
For Henzie, it also needs to check the Spell Mana Value because for X Creature Spells

so you can only Blitz them when X is high enough

Reply:
Yes, that wasn't happening at all before. The exclusive restriction wasn't defined in the right place. I moved it to the `Affected` param, and now it actually does check it (i.e. `Affected Creature.YouCtrl+cmcGE4`).

Reply:
I checked and I've put together a functionality matrix of expected behaviours across the `master` branch and this one (the fork).

| Expected Behaviour                                                                 | master | fork |
|------------------------------------------------------------------------------------|:------:|:----:|
| Standard card, alternative cost printed on it, can cast from all legal locations   |    ✔️    |   ✔️  |
| Standard card, alternative cost _granted to_ it, can cast from all legal locations |        |   ✔️  |
| MDFC Front, alternative cost printed on it, can cast from all legal locations      | ✔️      | ✔️    |
| MDFC Back, alternative cost printed on it, can cast from all legal locations       | ✔️      | ✔️    |
| MDFC Front, alternative cost _granted to_ it, can cast from all legal locations    |        | ✔️    |
| MDFC Back, alternative cost _granted to_ it, can cast from all legal locations     |        |      |
| Toolbox's ability correctly applies to creatures with mana value 4 or greater      |        | ✔️    |

The main takeaway is that the fork doesn't provide "MDFC Back, alternative cost _granted to_ it, can cast from all legal locations".

I can find a solution and add it to this PR, or I can address it in a separate one. It's your call. 👍

Reply:
The problem are X Creatures

Like Endless One with ManaCost =X

You allow to X them, but only if you choose X>=4

That's why it needs to apply to all creature Spells first, and is later checked again when you choose X and the spell is put on the Stack

In any case you choose X <=3 then you can't cast the spell and the action is canceled 


The MDFC where working with Blitz before, you need to test them different to prowl
So you should fix the MDFC in this MR too

Reply:
> The MDFC where working with Blitz before, you need to test them different to prowl
So you should fix the MDFC in this MR too

Will do!

> The problem are X Creatures
> 
> Like Endless One with ManaCost =X
> 
> You allow to X them, but only if you choose X>=4
> 
> That's why it needs to apply to all creature Spells first, and is later checked again when you choose X and the spell is put on the Stack
> 
> In any case you choose X <=3 then you can't cast the spell and the action is canceled

Was there a rules errata that lets Henzie work this way? Cursory searches online indicate that Henzie works the way you describe, but I can't find any official sources for it. I want to make sure I code this properly so having actual rules text for it would be ideal.

Thanks so much for reviewing this PR btw, I appreciate the thoroughness!

Reply:
Hey there!

Thanks for trying to work on the engine side of Forge.

Unfortunately you picked a pretty complex area as your first project :/

We haven't been happy with the MayPlay / alternative cost handling for quite some time, but it has also proven rather difficult to fix smaller aspects of it without breaking something else at the same time...(so it has been left alone mostly for a bigger refactor)

For your X interaction it's this:
> 202.3e. When calculating the mana value of an object with an {X} in its mana cost, X is treated as 0 while the object is not on the stack, and X is treated as the number chosen for it while the object is on the stack.

Reply:
> Unfortunately you picked a pretty complex area as your first project :/
Just my luck! 😆 

> We haven't been happy with the MayPlay / alternative cost handling for quite some time, but it has also proven rather difficult to fix smaller aspects of it without breaking something else at the same time...(so it has been left alone mostly for a bigger refactor)
Right, fair enough. It is turning out to be much hairier than anticipated. Would it be better to just shelve this in lieu of a more comprehensive refactor later, then?

> For your X interaction it's this:
>
> 202.3e. When calculating the mana value of an object with an {X} in its mana cost, X is treated as 0 while the object is not on the stack, and X is treated as the number chosen for it while the object is on the stack.
Thank you! I was very confused but this makes sense. I'm obviously very new to the codebase so I'm not clear on how Forge lines things up relative to the "proposal to cast a spell". When the user clicks a card in their hand, for example, Forge appears to get all costs and alternative costs at that point, and present them as a list for the user. For X cost creatures in Toolbox's case, should it present the Blitz option the way it normally would, but with X in the cost, then prompt for X, then check the total mana value of the spell on the stack and either cancel/prompt for mana?

Reply:
> Right, fair enough. It is turning out to be much hairier than anticipated. Would it be better to just shelve this in lieu of a more comprehensive refactor later, then?

It's your choice I guess. I'm not saying you can't try to tackle this and splitting it across consecutive PR is also sometimes easier. Just don't want you getting frustrated by this too quickly ;)

You noticed correctly that spell casting is one of the parts where we're not super accurate in the order according to the CR (sometimes it's also different for AI vs Human PlayerController). I suppose we could have less trouble in that regard if we would basically allow any card to be moved to the stack first as outlined by 601.2a. But then again this might end up too confusing GUI wise, which is one of the reasons we're doing some of these choices in sort of a "look-ahead" fashion. Another one would be this:
> 601.3a. If an effect prohibits a player from casting a spell with certain qualities, that player may consider any choices to be made during that spell's proposal that may cause those qualities to change. If any such choices could cause that effect to no longer prohibit that player from casting that spell, the player may begin to cast the spell, ignoring the effect.

So yea, that gets messy pretty quickly... 🤹‍♂️

Reply:
Ah, geez. It all makes sense now. That clears everything up, thanks @Hanmac.

I saw the comment in `getAlternativeCosts` that said `// TODO with mana value 4 or greater has blitz.` and that, coupled with the fact that the Blitz cost was appearing on the list of casting options when clicking a card, led me to believe that the behaviour hadn't been implemented yet. If I'd tried to actually click the Blitz option, I would have seen that it didn't work for mana values < 4.

It didn't help that I didn't know about rule `202.3e`, either (cheers, @tool4ever!).

Thanks for your patience, @Hanmac. I've been very obtuse.

So, to summarise:
- On `master`, static abilities that grant alternative costs work for normal cards and both faces of MDFCs in the hand, but not when able to cast from the top of the library or the command zone, etc.
- On the fork, static abilities that grant alternative costs work for normal cards and the front face of MDFCs for all zones they could normally be cast from, but not the back face of MDFCs, and it also breaks Toolbox for X costs (sorry about that 😩 )

I thought that moving the static ability checking code back to where it came from and duplicating the loop over `source.mayPlay(activator)`  would do the trick for the back face of MDFCs, but it doesn't seem to have worked. I'd prefer to find a way to enable the extra `CardPlayOption`s that minimises disruption, but I'll need to do some more investigation.

Thanks again, both of you, for taking the time to educate me on the codebase!

Reply:
> It's your choice I guess. I'm not saying you can't try to tackle this and splitting it across consecutive PR is also sometimes easier. Just don't want you getting frustrated by this too quickly ;)

Oh I'm happy to keep thrashing away at this. I've been a software engineer for over a decade now so I'm no stranger to frustration 😆

Thanks for the warm welcome. I'm looking forward to learning more about how Forge does things. I might take a look at some of the beginner friendly issues to get a bit more orientation, as well.

> So yea, that gets messy pretty quickly... 🤹‍♂️
Yes, absolutely. Those two rules, if implemented faithfully, would turn Forge into a Christopher Nolan film. 😬